### DISCUSSION TITLE
Are Inserts in Clickhouse Cloud Atomic? Partial Inserts Possible for ReplicatedMergeTree?

### DISCUSSION THREAD

**Discussion by kuro337** [ClickHouse/ClickHouse#46425] [Q&A]
Hello folks - I had a question. I am using Clickhouse Cloud and really enjoying the speed and performance. If anyone could assist with this question , would be highly appreciated!

I see in Clickhouse Docs, it is mentioned Inserts are ATOMIC in nature for a MergeTree for blocks up to `max_insert_block_size`.

![image](https://user-images.githubusercontent.com/65412787/218890016-cc9577e9-61b1-48db-86e1-78e4fe4ee44d.png)

The Engine Type is a `ReplicatedMergeTree`. 

I have Parquet files with several rows and I am looking to figure out whether or not INSERTS are atomic in nature or if partial inserts are possible. Can provide more info. 

According to this , need to design logic. 

Thank you for any help!

**Comment by den-crane** [comment]
https://kb.altinity.com/altinity-kb-queries-and-syntax/atomic-insert/

>I have Parquet files with several rows

There is a difference how you ingest Parquet, if you use clickhouse-client, it parses Parquet, if you use HTTP/curl then clickhouse-server parses Parquet.

**Reply by kuro337** [reply]
Thank you for pointing me towards those docs. I am currently creating a table from S3 where the parquet files are using something like this : 

```
CREATE OR REPLACE TABLE db.tablename (
  `date` datetime,
  ...fields,
) ENGINE S3(
   's3::/*', 
   'ak',
   'sk',
   'Parquet'
);
```

Then inserting using 

```
INSERT INTO CH_ReplicatedMergeTree (
 `fields`...
)
SELECT 
 `fields`...
FROM s3_table;
```

**Reply by den-crane** [reply]
This insert is not atomic.

**Reply by kuro337** [reply]
Is there a way I can ensure atomic inserts in a ReplicatedMergeTree table?

I ideally want to use clickhouse-go to insert data programatically from a Go app. (Or if there is a way to insert data from the Clickhouse Cloud UI that works as well)

**Reply by den-crane** [reply]
No way in general. Only some hacks like attach partition, but still it provides atomicity only on partition level.
Wait till transactions are implemented.

**Reply by UnamedRus** [reply]
Issue about improving documentation https://github.com/ClickHouse/clickhouse-docs/issues/444
